[BEAM-2844] Introduces PCollectionView.get()

jkff opened the pull request:
https://issues.apache.org/jira/browse/BEAM-2844

Supports it for DoFn in the various DoFnRunner's and DoFnTester. Support in CombineFn remains.

Introduces a category tag UsesImplicitSideInputs, currently excluded in Dataflow runner - that will be fixed in a follow-up PR after service-side changes, and the category will be removed.

R: @lukecwik

jkff commented:
retest this please

jkff commented:
Additional R: @kennknowles 

Note that `@SideInput` parameters are not in scope for this PR. My plan is:
1) Explicit wiring, but access via .get() - in DoFn: this PR and followup PRs after Dataflow worker changes
2) Same for CombineFn
3) Support for `@SideInput` fields and parameters for automatic wiring
4) somewhere in parallel with the above: simplification of APIs that used to require wiring but now wouldn't (e.g. deprecating CombineWithContext, DynamicDestinations.getSideInputs() etc)

jkff commented:
Run Dataflow ValidatesRunner

jkff commented:
Run Flink ValidatesRunner

jkff commented:
Run Spark ValidatesRunner

jkff commented:
Run Dataflow ValidatesRunner

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/DoFnTesterTest.java:
```diff
@@ -405,6 +425,18 @@ public void processElement(ProcessContext c) throws Exception {
     }
   }
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
Add a failure test showing that you can't access the side input during startBundle/finishBundle.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:
```diff
@@ -529,4 +534,49 @@ public String toString() {
+    <T> T sideInput(PCollectionView<T> view);
+  }
+
+  private static ThreadLocal<SideInputContext> sideInputContextThreadLocal = new ThreadLocal<>();
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
We found out that thread local is pretty slow compared to a ConcurrentHashMap with the key as the current thread. If your worried about leaking information you can always use a guava cache with weak keys.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:
```diff
@@ -529,4 +534,49 @@ public String toString() {
+
+  /** Accessor for the value of a side input in the current context. */
+  @Internal
+  public interface SideInputContext {
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
It makes me sad that this isn't in runners core due to things like DoFnTester.

lukecwik commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java:
```diff
@@ -172,9 +175,10 @@ public void onTimer(
     // This can contain user code. Wrap it in case it throws an exception.
-    try {
-      invoker.invokeProcessElement(new DoFnProcessContext(elem));
+    try (ScopedSideInputContext ignored = setCurrentSideInputContext(context)) {
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
nit: Technically not ignored since your relying on the try with resources to do work for you.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:
```diff
@@ -529,4 +534,49 @@ public String toString() {
+    SideInputContext context = sideInputContextThreadLocal.get();
+    if (context == null) {
+      throw new IllegalStateException(
+          "Current thread is not in a context where side inputs are available");
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
Should we specifically mention processElement?

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SplittableDoFnTest.java:
```diff
@@ -294,6 +294,41 @@ public void testSideInput() throws Exception {
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesSplittableParDo.class})
```
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
Missing UsesImplicitSideInputs

lukecwik commented:
Please also open a JIRA about adding a get method to StateSpec which provides the Bag/Value/...State out of it and deprecating the other contexts.

jkff commented:
On your high-level comment: I decided to go for option 3 but improve error reporting - see changes in PCollectionViews. I can't think of a reliable way to find which PCollectionView's inside a closure will actually be accessed; and there might be PCollectionView's captured not in the closure itself but transitively through its member variables etc. - this also rules out option 2. I think good error messages are the only feasible way.

I added error checking for ParDo test for accessing from StartBundle/FinishBundle.

Deprecation of the methods you indicated will follow as part of the parent JIRA, but shouldn't be done in this PR until Dataflow runner supports UsesImplicitSideInputs.

Note that, since I changed PCollectionView interface, I had to add a serialVersionUID matching the old value to SimplePCollectionView - otherwise mismatch of runners-core-construction jars between the pipeline and the worker image caused it to fail with NonSerializableException's about this class.

Filed https://issues.apache.org/jira/browse/BEAM-2950 about state.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java:
```diff
@@ -315,17 +271,28 @@ public void testNoTestPipelineUsedNoAnnotation() {}
+
+      @BeforeClass
+      public static void beforeClass() {
+        oldUseDefaultDummyRunner =
```
Please use the JUnit test rule RestoreSystemProperties instead of a specialized implementation.

lukecwik commented on runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/RunnerPCollectionView.java:
```diff
@@ -35,6 +37,8 @@
 
 /** A {@link PCollectionView} created from the components of a {@link SideInput}. */
 class RunnerPCollectionView<T> extends PValueBase implements PCollectionView<T> {
+  public static final long serialVersionUID = 2465466441736952633L;
```
make serialVersionUID private

The JVM reads this property directly and is not meant to be public.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:
```diff
@@ -528,5 +535,102 @@ public String toString() {
+          transform);
+    }
+
+    private void writeObject(ObjectOutputStream os)
```
Why do you need to implement Java's special serialization hook methods (writeObject/readObject)?

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:
```diff
@@ -564,6 +565,111 @@ public void testParDoWithSideInputs() {
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesImplicitSideInputs.class})
```
To make this a ValidatesRunner test for a failure scenario catch and convert the exception into a stack trace and match against the contents of the string using PAssert. There is no expectation that a test pipeline runner can accurately rethrow the failure reason to the process that submitted the job.

Ditto on the finishBundle test as well.

kennknowles commented:
Please do not conflate this to also be the model for `StateSpec`.

lukecwik commented on sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:
```diff
@@ -272,7 +275,7 @@ public void startBundle() {
   public void processElement(WindowedValue<InputT> elem) {
     currentElement = elem;
-    try {
+    try (ScopedSideInputContext ignored = setCurrentSideInputContext(processBundleContext)) {
```
Scope the side input access from the end of startBundle to the beginning of finishBundle so that we aren't updating a thread local on each process element call.

jkff commented:
I'm going to close this because it's no longer likely to be the right way to go.
See http://s.apache.org/context-fn
